File: rmw_int2dds_validation/validation/qos/liveliness/rclpy/qos_liveliness_pub.py
#!/usr/bin/env python3

# Copyright 2026 Int2DDS Project
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import sys
import time

import rclpy
from rclpy.duration import Duration
from rclpy.executors import SingleThreadedExecutor
from rclpy.node import Node
from rclpy.qos import (DurabilityPolicy, HistoryPolicy, LivelinessPolicy, QoSProfile,
                       ReliabilityPolicy)
from std_msgs.msg import String


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument('--topic', required=True)
    parser.add_argument('--liveliness', default='automatic')
    parser.add_argument('--lease-sec', type=float, required=True)
    parser.add_argument('--seed-prefix', default='alive')
    parser.add_argument('--seed-delay', type=float, default=0.1)
    parser.add_argument('--repeat-count', type=int, default=1)
    parser.add_argument('--repeat-period', type=float, default=0.5)
    parser.add_argument('--assert-period', type=float, default=0.0)
    parser.add_argument('--keep-alive', type=float, default=2.0)
    parser.add_argument('--wait-for-subscriber-count', type=int, default=0)
    parser.add_argument('--wait-timeout', type=float, default=5.0)
    return parser.parse_args()


def parse_liveliness(value: str) -> LivelinessPolicy:
    if value == 'manual_by_topic':
        return LivelinessPolicy.MANUAL_BY_TOPIC
    return LivelinessPolicy.AUTOMATIC


class LivelinessPublisher(Node):
    def __init__(self, args: argparse.Namespace) -> None:
        super().__init__('qos_liveliness_pub', start_parameter_services=False, enable_rosout=False)
        self.args = args
        self.start_time = time.time()
        self.done = False
        self.publish_count = 0
        self.assert_count = 0

        qos = QoSProfile(
            history=HistoryPolicy.KEEP_LAST,
            depth=10,
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE,
            liveliness=parse_liveliness(args.liveliness),
            liveliness_lease_duration=Duration(seconds=args.lease_sec),
        )

        self.publisher = self.create_publisher(String, args.topic, qos)
        self.observe_timer = self.create_timer(0.5, self._observe)
        self.publish_timer = self.create_timer(max(args.seed_delay, 0.01), self._publish_once)
        self.assert_timer = None
        if args.assert_period > 0.0:
            self.assert_timer = self.create_timer(
                max(args.assert_period, 0.01), self._assert_liveliness)
        self.shutdown_timer = self.create_timer(
            max(args.seed_delay + args.keep_alive, 0.1), self._finish
        )
        self.get_logger().info(
            'publisher ready: '
            f'topic={args.topic}, liveliness={args.liveliness}, lease={args.lease_sec}s, '
            f'assert_period={args.assert_period}s'
        )

    def _observe(self) -> None:
        self.get_logger().info(
            f'subscriber_count observation: {self.count_subscribers(self.args.topic)}'
        )

    def _publish_once(self) -> None:
        if self.args.wait_for_subscriber_count > 0:
            current_count = self.count_subscribers(self.args.topic)
            if current_count < self.args.wait_for_subscriber_count:
                if time.time() - self.start_time > self.args.wait_timeout:
                    self.publish_timer.cancel()
                    self.observe_timer.cancel()
                    self.shutdown_timer.cancel()
                    self.done = True
                    self.get_logger().error(
                        'publisher wait-for-match failed: '
                        f'expected={self.args.wait_for_subscriber_count}, current={current_count}'
                    )
                return

        msg = String()
        msg.data = f'{self.args.seed_prefix}-{self.publish_count}'
        self.publisher.publish(msg)
        self.get_logger().info(f'published: {msg.data}')
        self.publish_count += 1

        if self.publish_count >= self.args.repeat_count:
            self.publish_timer.cancel()
        else:
            self.publish_timer.timer_period_ns = int(self.args.repeat_period * 1e9)

    def _assert_liveliness(self) -> None:
        if self.publish_count == 0:
            return
        self.publisher.assert_liveliness()
        self.assert_count += 1
        self.get_logger().info(f'asserted liveliness #{self.assert_count}')

    def _finish(self) -> None:
        self.shutdown_timer.cancel()
        self.observe_timer.cancel()
        if self.assert_timer is not None:
            self.assert_timer.cancel()
        self.done = True
        self.get_logger().info('publisher keep-alive finished')


def main() -> int:
    args = parse_args()
    rclpy.init()
    node = LivelinessPublisher(args)
    executor = SingleThreadedExecutor()
    executor.add_node(node)

    try:
        while rclpy.ok() and not node.done:
            executor.spin_once(timeout_sec=0.1)
    finally:
        executor.remove_node(node)
        node.destroy_node()
        executor.shutdown(timeout_sec=1.0)
        rclpy.shutdown()
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: rmw_int2dds_validation/validation/qos/liveliness/rclcpp/qos_liveliness_pub.cpp
// Copyright 2026 Int2DDS Project
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <chrono>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

namespace
{

rclcpp::LivelinessPolicy parse_liveliness(const std::string & value)
{
  if (value == "manual_by_topic") {
    return rclcpp::LivelinessPolicy::ManualByTopic;
  }
  return rclcpp::LivelinessPolicy::Automatic;
}

class LivelinessPublisherCpp : public rclcpp::Node
{
public:
  LivelinessPublisherCpp(
    const std::string & topic,
    const std::string & liveliness,
    double lease_sec,
    const std::string & seed_prefix,
    double seed_delay_sec,
    int repeat_count,
    double repeat_period_sec,
    double assert_period_sec,
    double keep_alive_sec,
    int wait_for_subscription_count,
    double wait_timeout_sec)
  : Node("qos_liveliness_pub_cpp",
      rclcpp::NodeOptions().start_parameter_services(false).enable_rosout(false)),
    topic_(topic),
    seed_prefix_(seed_prefix),
    repeat_count_(repeat_count),
    repeat_period_sec_(repeat_period_sec),
    assert_period_sec_(assert_period_sec),
    wait_for_subscription_count_(wait_for_subscription_count),
    wait_timeout_sec_(wait_timeout_sec)
  {
    rclcpp::QoS qos(rclcpp::KeepLast(10));
    qos.reliability(rclcpp::ReliabilityPolicy::Reliable);
    qos.durability(rclcpp::DurabilityPolicy::Volatile);
    qos.liveliness(parse_liveliness(liveliness));
    qos.liveliness_lease_duration(rclcpp::Duration::from_seconds(lease_sec));

    publisher_ = create_publisher<std_msgs::msg::String>(topic, qos);
    observe_timer_ = create_wall_timer(500ms, [this]() {
          RCLCPP_INFO(get_logger(), "subscriber_count observation: %zu", count_subscribers(topic_));
    });
    publish_timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::duration<double>(std::max(seed_delay_sec, 0.01))),
      [this]() {publish_once();});
    shutdown_timer_ = create_wall_timer(
      std::chrono::duration_cast<std::chrono::nanoseconds>(
        std::chrono::duration<double>(std::max(seed_delay_sec + keep_alive_sec, 0.1))),
      [this]() {
        shutdown_timer_->cancel();
        observe_timer_->cancel();
        if (assert_timer_) {
          assert_timer_->cancel();
        }
        done_ = true;
        RCLCPP_INFO(get_logger(), "publisher keep-alive finished");
      });
    if (assert_period_sec_ > 0.0) {
      assert_timer_ = create_wall_timer(
        std::chrono::duration_cast<std::chrono::nanoseconds>(
          std::chrono::duration<double>(std::max(assert_period_sec_, 0.01))),
        [this]() {assert_once();});
    }

    start_time_ = now();
    RCLCPP_INFO(
      get_logger(),
      "publisher ready: topic=%s, liveliness=%s, lease=%.3fs, assert_period=%.3fs",
      topic.c_str(),
      liveliness.c_str(),
      lease_sec,
      assert_period_sec_);
  }

  bool done() const
  {
    return done_;
  }

private:
  void publish_once()
  {
    if (wait_for_subscription_count_ > 0) {
      const auto current_count = count_subscribers(topic_);
      if (static_cast<int>(current_count) < wait_for_subscription_count_) {
        if ((now() - start_time_).seconds() > wait_timeout_sec_) {
          publish_timer_->cancel();
          observe_timer_->cancel();
          shutdown_timer_->cancel();
          done_ = true;
          RCLCPP_ERROR(
            get_logger(),
            "publisher wait-for-match failed: expected=%d, current=%zu",
            wait_for_subscription_count_,
            current_count);
        }
        return;
      }
    }

    std_msgs::msg::String msg;
    msg.data = seed_prefix_ + "-" + std::to_string(publish_count_);
    publisher_->publish(msg);
    ++publish_count_;
    RCLCPP_INFO(get_logger(), "published: %s", msg.data.c_str());

    if (publish_count_ >= repeat_count_) {
      publish_timer_->cancel();
    } else {
      publish_timer_->reset();
      publish_timer_->cancel();
      publish_timer_ = create_wall_timer(
        std::chrono::duration_cast<std::chrono::nanoseconds>(
          std::chrono::duration<double>(repeat_period_sec_)),
        [this]() {publish_once();});
    }
  }

  void assert_once()
  {
    if (publish_count_ == 0) {
      return;
    }
    if (!publisher_->assert_liveliness()) {
      RCLCPP_ERROR(get_logger(), "assert_liveliness failed");
      return;
    }
    ++assert_count_;
    RCLCPP_INFO(get_logger(), "asserted liveliness #%d", assert_count_);
  }

  std::string topic_;
  std::string seed_prefix_;
  int repeat_count_;
  double repeat_period_sec_;
  double assert_period_sec_;
  int wait_for_subscription_count_;
  double wait_timeout_sec_;
  bool done_{false};
  int publish_count_{0};
  int assert_count_{0};
  rclcpp::Time start_time_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
  rclcpp::TimerBase::SharedPtr observe_timer_;
  rclcpp::TimerBase::SharedPtr publish_timer_;
  rclcpp::TimerBase::SharedPtr assert_timer_;
  rclcpp::TimerBase::SharedPtr shutdown_timer_;
};

}  // namespace

int main(int argc, char ** argv)
{
  std::string topic;
  std::string liveliness = "automatic";
  double lease_sec = 0.5;
  std::string seed_prefix = "alive";
  double seed_delay_sec = 0.1;
  int repeat_count = 1;
  double repeat_period_sec = 0.5;
  double assert_period_sec = 0.0;
  double keep_alive_sec = 2.0;
  int wait_for_subscription_count = 0;
  double wait_timeout_sec = 5.0;

  for (int i = 1; i < argc; ++i) {
    const std::string arg(argv[i]);
    auto next_value = [&](const char * name) -> std::string {
        if (i + 1 >= argc) {
          throw std::runtime_error(std::string("missing value for ") + name);
        }
        return argv[++i];
      };
    if (arg == "--topic") {
      topic = next_value("--topic");
    } else if (arg == "--liveliness") {
      liveliness = next_value("--liveliness");
    } else if (arg == "--lease-sec") {
      lease_sec = std::stod(next_value("--lease-sec"));
    } else if (arg == "--seed-prefix") {
      seed_prefix = next_value("--seed-prefix");
    } else if (arg == "--seed-delay") {
      seed_delay_sec = std::stod(next_value("--seed-delay"));
    } else if (arg == "--repeat-count") {
      repeat_count = std::stoi(next_value("--repeat-count"));
    } else if (arg == "--repeat-period") {
      repeat_period_sec = std::stod(next_value("--repeat-period"));
    } else if (arg == "--assert-period") {
      assert_period_sec = std::stod(next_value("--assert-period"));
    } else if (arg == "--keep-alive") {
      keep_alive_sec = std::stod(next_value("--keep-alive"));
    } else if (arg == "--wait-for-subscriber-count") {
      wait_for_subscription_count = std::stoi(next_value("--wait-for-subscriber-count"));
    } else if (arg == "--wait-timeout") {
      wait_timeout_sec = std::stod(next_value("--wait-timeout"));
    } else {
      throw std::runtime_error("unknown argument: " + arg);
    }
  }

  if (topic.empty()) {
    throw std::runtime_error("--topic is required");
  }

  rclcpp::init(argc, argv);
  auto node = std::make_shared<LivelinessPublisherCpp>(
    topic,
    liveliness,
    lease_sec,
    seed_prefix,
    seed_delay_sec,
    repeat_count,
    repeat_period_sec,
    assert_period_sec,
    keep_alive_sec,
    wait_for_subscription_count,
    wait_timeout_sec);

  rclcpp::executors::SingleThreadedExecutor executor;
  executor.add_node(node);
  while (rclcpp::ok() && !node->done()) {
    executor.spin_once(100ms);
  }
  executor.remove_node(node);
  rclcpp::shutdown();
  return 0;
}
